dds/idl/marshal_generator.cpp

```diff
@@ -2989,21 +2994,41 @@ bool marshal_generator::gen_union(AST_Union* node, UTL_ScopedName* name,
     serialized_size.addArg("uni", "const " + cxx + "&");
     serialized_size.endArgs();
 
+    const string align = getAlignment(discriminator);
+    if (!align.empty()) {
+      be_global->impl_ << "  encoding.align(size, " << align << ");\n";
+    }
+
     std::vector<string> code;
     code.push_back("serialized_size_delimiter(encoding, size);");
     generate_dheader_code(code, not_final, false);
 
-    const string align = getAlignment(discriminator);
-    if (!align.empty()) {
-      be_global->impl_ << "  encoding.align(size, " << align << ");\n";
+    if (may_be_parameter_list) {
+      be_global->impl_ <<
+        "  size_t mutable_running_total = 0;\n";
+      be_global->impl_ <<
+        "  serialized_size_parameter_id(encoding, size, mutable_running_total);\n";
+      be_global->impl_ <<
+        "  size = mutable_running_total;\n";
```

Comment:
```suggestion
      be_global->impl_ <<
        "  size_t mutable_running_total = 0;\n"
        "  serialized_size_parameter_id(encoding, size, mutable_running_total);\n";
```

I don't think you should be messing with `size` and `mutable_running_total` like this. The `serialized_size_parameter_id*` functions will modify them as needed.

Reply:
It did not work when I tried it that way, as it looks like `size` is reset to 0 before serialized_size_parameter_id returns.
Of course, I may be missing something.

ACE_INLINE
void serialized_size_parameter_id(
  const Encoding& encoding, size_t& size, size_t& running_size)
{
  const Encoding::XcdrVersion xcdr = encoding.xcdr_version();
  if (xcdr == Encoding::XCDR_VERSION_1) {
    encoding.align(size, xcdr1_pid_alignment);
    size += uint16_cdr_size * 2;
    // TODO(iguessthislldo): Extended PID

    // Save and Zero Size to Reset the Alignment
    running_size += size;
    size = 0;
  } else if (xcdr == Encoding::XCDR_VERSION_2) {
    if (running_size != 0 && size != 1 && size != 2 && size != 4 && size != 8) {
      size += uint32_cdr_size; // nextint
    }
    encoding.align(size, uint32_cdr_size);
    size += uint32_cdr_size; // emheader
    running_size += size;
    size = 0;
  }
}

Reply:
Yes that's on purpose.

Reply:
Its keeps the running total separate, because it needs just the size of the fields when writing the PID/EMHEADER

Reply:
I meant "calculating the size of", not "writing". Then `serialized_size_list_end_parameter_id` is supposed to make `size` proper again at the end of the function.

Reply:
OK - thank you.